### PR TITLE
Enforce canonical release identity across GitHub and GHCR

### PR DESCRIPTION
**Pull request by wallstop**
## What changed

- resolve one canonical `X.Y.Z` / annotated `vX.Y.Z` / source-commit identity before release publication
- invoke the reusable multi-architecture container workflow directly from the release path
- publish and verify matching `vX.Y.Z`, `X.Y.Z`, rolling `X.Y`, and immutable `sha-*` image tags
- make crate, GHCR, tag, and GitHub Release retries fail closed and idempotent
- add per-platform OCI label verification, digest-bearing Release notes, a historical-tag backfill path, and a public release runbook
- add data-driven workflow-policy regression coverage for manual releases, direct tag pushes, backfills, and partial retries
- keep rebased force-push hook validation sub-second by excluding commits already reachable from the target remote and fast-gating semantic scans

## Why

The manual release workflow created its Git tag with `GITHUB_TOKEN`. GitHub does not start another workflow from that token-generated tag push, so a crate and GitHub Release could publish without versioned GHCR tags. The live `v0.4.0` audit confirms the gap: the annotated tag and crates.io package both identify commit `50b28a9…`, but `v0.4.0` and `0.4.0` are absent from GHCR.

## Compatibility and operations

Main-branch `latest` publication and direct human annotated tag pushes remain supported. Historical backfills rebuild from the annotated tag and preserve an older immutable SHA image instead of aliasing `latest` or moving existing bytes. After merge, dispatch **Docker Publish** with `release_tag=v0.4.0` to close the observed historical gap.

## Validation

- actionlint and shellcheck
- targeted clippy for `ci_config_tests`
- all 277 CI configuration tests (276 passed, one intentionally ignored expensive nested compile)
- all 14 documentation consistency policy/script tests
- CI config, tooling parity, workflow hygiene, markdown, LLM, and hook policy checks
- final SHA `940ac28`: all 11 applicable workflows green, both automated reviewers clean, zero unresolved threads
- enforced outgoing pre-push profile: 637 ms; actual push: 796 ms
- live GHCR verifier positive run against `latest`
- live GHCR verifier negative run proving the historical `sha-50b28a9` OCI-version mismatch

---

> [!NOTE]
> **High Risk**
> Changes the production release and GHCR publication pipeline with strict identity checks and tag immutability; a regression could block releases or mis-tag images, though retries are designed to fail closed rather than overwrite bytes.
> 
> **Overview**
> Unifies release and container publishing so **one version, annotated tag, and commit SHA** must agree across Cargo, CHANGELOG, crates.io, GitHub Release, and GHCR before anything ships.
> 
> **Release workflow** adds `resolve-release` and `ensure-tag` (immutable annotated tags that are never moved), calls **Docker Publish** as a reusable workflow with explicit `source_revision` (so `GITHUB_TOKEN` tag pushes no longer skip container publication), idempotent **crates.io** retries keyed on published git SHA, and Release notes that record the verified multi-arch **manifest digest** and source revision.
> 
> **Docker Publish** becomes a shared entry point for main (`:latest` / `sha-*`), direct `v*` tag pushes, release workflow calls, and manual **historical backfills**; release builds can **reuse a verified digest** and only repair missing tag aliases, with `scripts/verify-release-image.sh` checking platforms and OCI labels.
> 
> Also adds **`docs/releasing.md`**, changelog notes, expanded **`ci_config_tests.rs`** workflow policy tests, and small **pre-commit/pre-push** hook optimizations for rebased force-pushes and faster policy scans.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 940ac283402dfce1914140489165d423de846db6. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by wallstop** [comment]
bugbot run

**Comment by wallstop** [comment]
bugbot run

**Comment by wallstop** [comment]
bugbot run

**Comment by wallstop** [comment]
bugbot run

**Comment by wallstop** [comment]
bugbot run

**Comment by wallstop** [comment]
@copilot review

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -52,17 +52,87 @@ on:
...
+            echo "ERROR: Workflow version ($version), Git tag ($tag), and Cargo.toml ($cargo_version) disagree." >&2
+            exit 1
+          fi
+          if [[ ! "$version" =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
+            echo "ERROR: Release version is not canonical semantic version text: $version" >&2
+            exit 1
+          fi
+          if ! grep -Fqx "## [${version}]" CHANGELOG.md; then
+            echo "ERROR: CHANGELOG.md has no exact ## [${version}] release section." >&2
+            exit 1
```
[the remark this answers is not in the document]
Comment (wallstop) [reply .github/workflows/release.yml:112]: Fixed in `3f13858`: the gate now accepts only `## [X.Y.Z]` or `## [X.Y.Z] - YYYY-MM-DD`, with the version escaped before regex matching. Both accepted forms and invalid trailing text were exercised locally.

File: scripts/hooks/pre-push.ps1 [hunk 2]
```diff
@@ -79,7 +79,11 @@ function Get-RevList {
         $remoteArg = if ([string]::IsNullOrWhiteSpace($RemoteName)) { "--remotes" } else { "--remotes=$RemoteName" }
         $result = Invoke-Native -FileName "git" -Arguments @("rev-list", $LocalSha, "--not", $remoteArg)
     } else {
-        $result = Invoke-Native -FileName "git" -Arguments @("rev-list", "$RemoteSha..$LocalSha")
+        # A force-push after rebasing can place commits from another remote branch
+        # outside RemoteSha..LocalSha. Those commits are already present on the
+        # target remote, so only inspect commits the push would newly introduce.
+        $remoteArg = if ([string]::IsNullOrWhiteSpace($RemoteName)) { "--remotes" } else { "--remotes=$RemoteName" }
+        $result = Invoke-Native -FileName "git" -Arguments @("rev-list", $LocalSha, "--not", $RemoteSha, $remoteArg)
```
[the remark this answers is not in the document]
Comment (wallstop) [reply scripts/hooks/pre-push.ps1:86]: Audited; no change needed. The hook’s scope is commits newly introduced to the target remote, matching the existing new-branch `--not --remotes=<remote>` behavior. A commit already reachable from `origin/main` is not new remote content and re-scanning it caused the measured rebase regression.

File: .github/workflows/docker-publish.yml [hunk 3]
```diff
@@ -98,25 +215,143 @@ jobs:
...
+          SHA_TAG: sha-${{ needs.resolve.outputs.short_sha }}
+          PUBLISH_SHA: ${{ steps.existing.outputs.publish_sha }}
+        run: |
+          set -euo pipefail
+          tags=("$RELEASE_TAG" "$RELEASE_VERSION")
+          if [ "$PUBLISH_SHA" != "false" ]; then
+            tags+=("$SHA_TAG")
+          fi
+          bash scripts/verify-release-image.sh \
+            "$IMAGE" "$SOURCE_REVISION" "$RELEASE_VERSION" "${tags[@]}"
```
[the remark this answers is not in the document]
Comment (wallstop) [reply .github/workflows/docker-publish.yml:353]: Fixed in `56af61a`: the verification tag set now includes `$MAJOR_MINOR` alongside the annotated tag, full semver tag, and applicable SHA tag.

File: scripts/hooks/pre-push.ps1 [hunk 4]
```diff
@@ -548,6 +557,13 @@ function Test-CommandTextForDirectScript {
         [System.Collections.Generic.List[string]]$Violations
     )
 
+    # Most lines in multiline run blocks cannot reference a repository script.
+    # Avoid the quote-aware shell tokenization path unless its only supported
+    # path prefixes are present.
+    if (-not $CommandText.Contains("scripts/")) {
+        return
+    }
```
[the remark this answers is not in the document]
Comment (wallstop) [reply scripts/hooks/pre-push.ps1:565]: Audited; no change needed. `.github/scripts/` contains the literal substring `scripts/`, so this fast gate passes those lines to `Normalize-LocalScriptToken`; the existing direct-invocation edge-case test remains green.

File: .github/workflows/docker-publish.yml [hunk 5]
```diff
@@ -55,28 +20,180 @@ on:
...
+            fi
+            if [[ ! "$release_version" =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
+              echo "ERROR: Release version is not canonical semantic version text: $release_version" >&2
+              exit 1
+            fi
+            if [ "$(git cat-file -t "refs/tags/${release_tag}" 2>/dev/null || true)" != "tag" ]; then
+              echo "ERROR: $release_tag must exist as an annotated Git tag before container publication." >&2
+              exit 1
+            fi
+            tag_revision=$(git rev-parse "refs/tags/${release_tag}^{commit}")
```
[the remark this answers is not in the document]
Comment (wallstop) [reply .github/workflows/docker-publish.yml:149]: Fixed in `800273e`: the Docker resolve job explicitly fetches `refs/tags/${release_tag}` from origin before validating the tag object type and target commit.

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -52,17 +52,88 @@ on:
...
+            echo "ERROR: CHANGELOG.md has no ## [${version}] release section, with or without a date." >&2
+            exit 1
+          fi
+
+          if [ "${{ github.event_name }}" = "push" ]; then
+            if [ "$(git cat-file -t "refs/tags/${tag}" 2>/dev/null || true)" != "tag" ]; then
+              echo "ERROR: Direct release tag $tag must be annotated, not lightweight." >&2
+              exit 1
+            fi
+            tag_revision=$(git rev-parse "refs/tags/${tag}^{commit}")
```
[the remark this answers is not in the document]
Comment (wallstop) [reply .github/workflows/release.yml:121]: Fixed in `800273e`: direct tag releases explicitly fetch `refs/tags/${tag}` from origin before `git cat-file` / `rev-parse` validation.

File: scripts/verify-release-image.sh [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    echo "ERROR: '$command' is required to verify a release image." >&2
+    exit 1
+  fi
+done
+
+if [[ ! "$expected_revision" =~ ^[0-9a-f]{40}$ ]]; then
+  echo "ERROR: expected source revision must be a full 40-character lowercase Git SHA." >&2
+  exit 1
+fi
+
```
[the remark this answers is not in the document]
Comment (wallstop) [reply scripts/verify-release-image.sh:39]: Fixed in `800273e`: the verifier now runs `docker buildx version` up front and emits a dedicated Buildx-required error.

**Comment by wallstop** [comment]
bugbot run

File: scripts/verify-release-image.sh [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+resolved_digest=""
+
+for tag in "${tags[@]}"; do
+  reference="${image}:${tag}"
+  inspect=$(docker buildx imagetools inspect "$reference")
+  digest=$(printf '%s\n' "$inspect" | sed -n 's/^Digest:[[:space:]]*//p' | head -n 1)
+  if [[ ! "$digest" =~ ^sha256:[0-9a-f]{64}$ ]]; then
+    echo "ERROR: could not resolve an OCI digest for $reference." >&2
+    exit 1
+  fi
```
[the remark this answers is not in the document]
Comment (wallstop) [reply scripts/verify-release-image.sh:54]: Fixed in `bb73b12`: tag, canonical-manifest, and per-platform Buildx inspections now capture stderr/stdout and report the exact failing reference before exiting.

File: .github/workflows/docker-publish.yml [hunk 9]
```diff
@@ -55,28 +20,181 @@ on:
...
     runs-on: ubuntu-latest
     timeout-minutes: 30
+    concurrency:
+      group: container-publish-${{ needs.resolve.outputs.publish_latest == 'true' && 'latest' || needs.resolve.outputs.source_revision }}
+      cancel-in-progress: false
+    permissions:
+      contents: read
+      packages: write
+    outputs:
+      digest: ${{ steps.verify.outputs.digest }}
```
[the remark this answers is not in the document]
Comment (wallstop) [reply .github/workflows/docker-publish.yml]: Fixed in `bb73b12`: the job output now uses `steps.verify.outputs.digest || steps.rolling-digest.outputs.digest`, covering both release and main-branch publication paths.

**Comment by wallstop** [comment]
bugbot run

File: .github/workflows/release.yml [hunk 10]
```diff
@@ -52,17 +52,89 @@ on:
 permissions:
   contents: write
   actions: read
+  packages: write
```
[the remark this answers is not in the document]
Comment (wallstop) [reply .github/workflows/release.yml]: Fixed in `940ac28`: workflow-level `packages: write` is removed, the reusable `publish-container` job retains the scoped grant, and a config regression test prevents broadening it again.
